tpl/template_funcs.go

```diff
@@ -1392,6 +1392,47 @@ func replace(a, b, c interface{}) (string, error) {
 	return strings.Replace(aStr, bStr, cStr, -1), nil
 }
 
+// partialCache represents a cache of partials protected by a mutex.
+type partialCache struct {
+	mu sync.RWMutex
+	p  map[string]template.HTML
+}
+
+// Get retrieves partial output from the cache based upon the partial name.
+// If the partial is not found in the cache, the partial is rendered and added
+// to the cache.
+func (c *partialCache) Get(key, name string, context interface{}) (p template.HTML) {
+	var ok bool
+
+	if p, ok = c.get(key); !ok {
+		p = c.set(key, name, context)
+	}
+
+	return p
+}
+
+func (c *partialCache) get(key string) (p template.HTML, ok bool) {
+	c.mu.RLock()
+	p, ok = c.p[key]
+	c.mu.RUnlock()
+	return
+}
+
+func (c *partialCache) set(key, name string, context interface{}) template.HTML {
+	c.mu.Lock()
+	p := partial(name, []interface{}{context}...)
+	c.p[key] = p
+	c.mu.Unlock()
+	return p
+}
+
+var cachedPartials = partialCache{p: make(map[string]template.HTML)}
+
+func partialCached(name string, context interface{}, variant ...string) template.HTML {
+	key := name + strings.Join(variant, "")
```

Comment:
This is nitpicky, but I have seen some fairly heavy use of partials lately, so this would be better:

```go
key := name
if len(variant) > 0 {
   key += ...
}